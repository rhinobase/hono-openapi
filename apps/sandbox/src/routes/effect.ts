import { Schema } from "effect";
import { Hono } from "hono";
import { describeRoute } from "hono-openapi";
import { validator as effectValidator, resolver } from "hono-openapi/effect";

const router = new Hono();

const nameValidation = Schema.Struct({
  name: Schema.String,
});

router.get(
  "/",
  describeRoute({
    description: "Say hello to the user",
    responses: {
      200: {
        description: "Successful greeting response",
        content: {
          "text/plain": {
            schema: resolver(Schema.String),
          },
        },
      },
    },
  }),
  effectValidator("query", nameValidation),
  (c) => {
    const query = c.req.valid("query");
    return c.text(`Hello ${query?.name ?? "Hono"}!`);
  },
);

router.post(
  "/",
  describeRoute({
    description: "Create a new user",
    responses: {
      200: {
        description: "Successful user creation response",
        content: {
          "text/plain": {
            schema: {
              type: "string",
              example: "Hello Steven! Your id is 123",
            },
          },
        },
      },
    },
  }),
  effectValidator("query", nameValidation),
  effectValidator(
    "json",
    Schema.Struct({
      id: Schema.Number,
    }),
  ),
  (c) => {
    const query = c.req.valid("query");
    const body = c.req.valid("json");

    return c.text(`Hello ${query.name}! Your id is ${body.id}`);
  },
);

export default router;
