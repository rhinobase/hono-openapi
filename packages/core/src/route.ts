import { HTTPException } from "hono/http-exception";
import type { Env, Input, MiddlewareHandler } from "hono/types";
import type { DescribeRouteOptions, OpenAPIRouteHandlerConfig } from "./types";
import { uniqueSymbol } from "./utils";

export function describeRoute<
  E extends Env = Env,
  P extends string = string,
  I extends Input = Input,
>(specs: DescribeRouteOptions<E, P, I>): MiddlewareHandler<E, P, I> {
  const middleware: MiddlewareHandler<E, P, I> = async (c, next) => {
    await next();

    if (specs.validateResponse && specs.responses) {
      const status = c.res.status;
      const contentType = c.res.headers.get("content-type");

      if (status && contentType) {
        const response = specs.responses[status];
        if (response && "content" in response && response.content) {
          const content = response.content[contentType];
          if (content?.schema && "validator" in content.schema) {
            try {
              await content.schema.validator(c.res.body);
            } catch (error) {
              throw new HTTPException(400, {
                message: "Response validation failed!",
              });
            }
          }
        }
      }
    }
  };

  return Object.assign(middleware, {
    [uniqueSymbol]: {
      resolver: async (config: OpenAPIRouteHandlerConfig) => {
        const docs = { ...specs };
        let components = {};

        if (docs.responses) {
          for (const key of Object.keys(docs.responses)) {
            const response = docs.responses[key];
            if (response && !("content" in response)) continue;

            for (const contentKey of Object.keys(response.content ?? {})) {
              const raw = response.content?.[contentKey];

              if (!raw) continue;

              if (raw.schema && "builder" in raw.schema) {
                const result = await raw.schema.builder(config);
                raw.schema = result.schema;
                if (result.components) {
                  components = {
                    ...components,
                    ...result.components,
                  };
                }
              }
            }
          }
        }

        return { docs, components };
      },
    },
  });
}
