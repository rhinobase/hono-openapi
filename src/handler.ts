import type { Context, Env, Hono } from "hono";
import type {
  BlankEnv,
  BlankInput,
  BlankSchema,
  Input,
  Schema,
} from "hono/types";
import type { OpenAPIV3_1 } from "openapi-types";
import type {
  DescribeRouteOptions,
  GenerateSpecOptions,
  ResolverReturnType,
  SanitizedGenerateSpecOptions,
} from "./types";
import {
  ALLOWED_METHODS,
  type AllowedMethods,
  registerSchemaPath,
  removeExcludedPaths,
  uniqueSymbol,
} from "./utils";

const DEFAULT_OPTIONS: Partial<GenerateSpecOptions> = {
  documentation: {},
  excludeStaticFile: true,
  exclude: [],
  excludeMethods: ["OPTIONS"],
  excludeTags: [],
};

type SpecContext = {
  components: Record<string, OpenAPIV3_1.SchemaObject>;
  options: SanitizedGenerateSpecOptions;
};

/**
 * Generate OpenAPI specs for the given Hono instance
 * @param hono Instance of Hono
 * @param options Options for generating OpenAPI specs
 * @param config Configuration for OpenAPI route handler
 * @param Context Route context for hiding routes
 * @returns OpenAPI specs
 */
export async function generateSpecs<
  E extends Env = BlankEnv,
  P extends string = string,
  I extends Input = BlankInput,
  S extends Schema = BlankSchema,
>(
  hono: Hono<E, S, P>,
  options = DEFAULT_OPTIONS,
  c?: Context<E, P, I>,
) {
  const ctx: SpecContext = {
    components: {},
    // @ts-expect-error
    options: {
      ...DEFAULT_OPTIONS,
      ...options,
    },
  };

  const _documentation = ctx.options.documentation ?? {};
  const schema = await generatePaths(hono, ctx);

  // Hide routes
  for (const path in schema) {
    for (const method in schema[path]) {
      const isHidden = getHiddenValue({
        valueOrFunc: schema[path][method]?.hide,
        method,
        path,
        c,
      });

      if (isHidden) {
        delete schema[path][method];
      }
    }
  }

  return {
    openapi: "3.1.0",
    ..._documentation,
    tags: _documentation.tags?.filter(
      (tag) => !ctx.options.excludeTags?.includes(tag?.name),
    ),
    info: {
      title: "Hono Documentation",
      description: "Development documentation",
      version: "0.0.0",
      ..._documentation.info,
    },
    paths: {
      ...removeExcludedPaths(schema, ctx),
      ..._documentation.paths,
    },
    components: {
      ..._documentation.components,
      schemas: {
        ...ctx.components,
        ..._documentation.components?.schemas,
      },
    },
  } satisfies OpenAPIV3_1.Document;
}

async function generatePaths<
  E extends Env = BlankEnv,
  P extends string = string,
  S extends Schema = BlankSchema,
>(
  hono: Hono<E, S, P>,
  ctx: SpecContext,
): Promise<OpenAPIV3_1.PathsObject> {
  const paths: OpenAPIV3_1.PathsObject = {};

  for (const route of hono.routes) {
    // Finding routes with uniqueSymbol
    if (!(uniqueSymbol in route.handler)) {
      // Include empty paths, if enabled
      if (ctx.options.includeEmptyPaths) {
        registerSchemaPath({
          route,
          paths,
        });
      }

      continue;
    }

    const routeMethod = route.method as AllowedMethods | "ALL";

    // All method acts like a middleware, so we can skip it
    if (routeMethod !== "ALL") {
      // Exclude methods
      if (ctx.options.excludeMethods?.includes(routeMethod)) {
        continue;
      }

      // Include only allowed methods
      if (!ALLOWED_METHODS.includes(routeMethod)) {
        continue;
      }
    }

    const middlewareHandler = route.handler[uniqueSymbol] as
      | ResolverReturnType
      | { schema: DescribeRouteOptions; components: undefined };

    const defaultOptionsForThisMethod = ctx.options.defaultOptions
      ?.[routeMethod];

    const { schema: routeSpecs, components = {} } =
      "schema" in middlewareHandler
        ? middlewareHandler
        : await middlewareHandler
          .toOpenAPISchema({
            ...defaultOptionsForThisMethod,
            components: ctx.components,
          });

    ctx.components = {
      ...ctx.components,
      ...components,
    };

    registerSchemaPath({
      route,
      specs: routeSpecs,
      paths,
    });
  }

  return paths;
}

function getHiddenValue(options: {
  valueOrFunc: DescribeRouteOptions["hide"];
  c?: Context;
  method: string;
  path: string;
}) {
  const { valueOrFunc, c, method, path } = options;

  if (valueOrFunc != null) {
    if (typeof valueOrFunc === "boolean") {
      return valueOrFunc;
    } else if (typeof valueOrFunc === "function") {
      if (c) {
        return valueOrFunc(c);
      } else {
        console.warn(
          `'c' is not defined, cannot evaluate hide function for ${method} ${path}`,
        );
      }
    }
  }

  return false;
}
