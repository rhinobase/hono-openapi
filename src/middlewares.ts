import { type Hook, sValidator } from "@hono/standard-validator";
import type { Env, Input, MiddlewareHandler, ValidationTargets } from "hono";
import type { StandardSchemaV1 } from "@standard-schema/spec";
import { toJsonSchema } from "@standard-community/standard-json";
import { toOpenAPISchema } from "@standard-community/standard-openapi";
import { uniqueSymbol } from "./utils";
import { DescribeRouteOptions } from "./types";

/**
 * Generate a resolver for a validation schema
 * @param schema Validation schema
 * @returns Resolver result
 */
export function resolver<Schema extends StandardSchemaV1>(schema: Schema) {
  return {
    vendor: schema["~standard"].vendor,
    validate: schema["~standard"].validate,
    toJSONSchema: (options?: Record<string, unknown>) =>
      toJsonSchema(schema, options),
    toOpenAPI: (options?: Record<string, unknown>) =>
      toOpenAPISchema(schema, options),
  };
}

type HasUndefined<T> = undefined extends T ? true : false;

/**
 * Create a validator middleware
 * @param target Target for validation
 * @param schema Validation schema
 * @param hook Hook for validation
 * @returns Middleware handler
 */
export function validator<
  Schema extends StandardSchemaV1,
  Target extends keyof ValidationTargets,
  E extends Env,
  P extends string,
  In = StandardSchemaV1.InferInput<Schema>,
  Out = StandardSchemaV1.InferOutput<Schema>,
  I extends Input = {
    in: HasUndefined<In> extends true ? {
        [K in Target]?: In extends ValidationTargets[K] ? In
          : { [K2 in keyof In]?: ValidationTargets[K][K2] };
      }
      : {
        [K in Target]: In extends ValidationTargets[K] ? In
          : { [K2 in keyof In]: ValidationTargets[K][K2] };
      };
    out: { [K in Target]: Out };
  },
  V extends I = I,
>(
  target: Target,
  schema: Schema,
  hook?: Hook<StandardSchemaV1.InferOutput<Schema>, E, P, Target>,
  options?: Record<string, unknown>,
): MiddlewareHandler<E, P, V> {
  const middleware = sValidator(target, schema, hook);

  // @ts-expect-error not typed well
  return Object.assign(middleware, {
    [uniqueSymbol]: {
      target,
      ...resolver(schema),
      options,
    },
  });
}

/**
 * Describe a route with OpenAPI specs.
 * @param specs Options for describing a route
 * @returns Middleware handler
 */
export function describeRoute(specs: DescribeRouteOptions): MiddlewareHandler {
  const middleware: MiddlewareHandler = async (_c, next) => {
    await next();
  };

  return Object.assign(middleware, {
    [uniqueSymbol]: {
      specs,
    },
  });
}
