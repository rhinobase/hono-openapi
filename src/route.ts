import { HTTPException } from "hono/http-exception";
import type { MiddlewareHandler } from "hono/types";
import type {
  ClientErrorStatusCode,
  ServerErrorStatusCode,
} from "hono/utils/http-status";
import type {
  DescribeRouteOptions,
  OpenAPIRouteHandlerConfig,
} from "./types.js";
import { uniqueSymbol } from "./utils.js";

/**
 * Describe a route with OpenAPI specs.
 * @param specs Options for describing a route
 * @returns Middleware handler
 */
export function describeRoute(specs: DescribeRouteOptions): MiddlewareHandler {
  const { validateResponse, ...docs } = specs;

  const middleware: MiddlewareHandler = async (c, next) => {
    await next();

    if (validateResponse && specs.responses) {
      const status = c.res.status;
      const contentType = c.res.headers.get("content-type");

      if (status && contentType) {
        const response = specs.responses[status];
        if (response && "content" in response && response.content) {
          const splitedContentType = contentType.split(";")[0];
          const content = response.content[splitedContentType];

          if (content?.schema && "validator" in content.schema) {
            try {
              let data: unknown;
              const clonedRes = c.res.clone();

              if (splitedContentType === "application/json") {
                data = await clonedRes.json();
              } else if (splitedContentType === "text/plain") {
                data = await clonedRes.text();
              }

              if (!data) throw new Error("No data to validate!");

              await content.schema.validator(data);
            } catch (error) {
              let httpExceptionOptions: {
                status: ClientErrorStatusCode | ServerErrorStatusCode;
                message: string;
              } = {
                status: 500,
                message: "Response validation failed!",
              };

              if (typeof validateResponse === "object") {
                httpExceptionOptions = {
                  ...httpExceptionOptions,
                  ...validateResponse,
                };
              }

              throw new HTTPException(httpExceptionOptions.status, {
                message: httpExceptionOptions.message,
                cause: error,
              });
            }
          }
        }
      }
    }
  };

  return Object.assign(middleware, {
    [uniqueSymbol]: {
      resolver: (
        config: OpenAPIRouteHandlerConfig,
        defaultOptions?: DescribeRouteOptions,
      ) => generateRouteSpecs(config, docs, defaultOptions),
    },
  });
}

/**
 * Generate OpenAPI specs for the given route
 * @param config Route handler configuration
 * @param docs Route description in OpenAPI specs
 * @param defaultOptions Default options for describing a route
 */
export async function generateRouteSpecs(
  config: OpenAPIRouteHandlerConfig,
  docs: DescribeRouteOptions,
  defaultOptions: DescribeRouteOptions = {},
) {
  let components = {};
  const tmp = {
    ...defaultOptions,
    ...docs,
    responses: {
      ...defaultOptions?.responses,
      ...docs.responses,
    },
  };

  if (tmp.responses) {
    for (const key of Object.keys(tmp.responses)) {
      const response = tmp.responses[key];

      if (!response || !("content" in response)) continue;

      for (const contentKey of Object.keys(response.content ?? {})) {
        const raw = response.content?.[contentKey];

        if (!raw) continue;

        if (raw.schema && "builder" in raw.schema) {
          const result = await raw.schema.builder(config);
          raw.schema = result.schema;
          if (result.components) {
            components = {
              ...components,
              ...result.components,
            };
          }
        }
      }
    }
  }

  return { docs: tmp, components };
}
