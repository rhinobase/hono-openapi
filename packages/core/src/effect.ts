import { effectValidator } from "@hono/effect-validator";
import { JSONSchema, Schema } from "effect";
import type {
  Env,
  Input as HonoInput,
  MiddlewareHandler,
  ValidationTargets,
} from "hono";
import convert from "./toOpenAPISchema.js";
import type {
  HasUndefined,
  OpenAPIRouteHandlerConfig,
  ResolverResult,
} from "./types.js";
import { generateValidatorDocs, uniqueSymbol } from "./utils.js";

/**
 * Generate a resolver for an Effect schema
 * @param schema Effect schema
 * @returns Resolver result
 */
export function resolver<Type, Context>(
  schema: Schema.Schema<Type, Context, never>,
): ResolverResult {
  return {
    builder: async (options?: OpenAPIRouteHandlerConfig) => ({
      schema: await convert(JSONSchema.make(schema)),
    }),
    validator: async (value) => {
      await Schema.decodeUnknownPromise(schema)(value);
    },
  };
}

/**
 * Create a validator middleware
 * @param target Target for validation
 * @param schema Effect schema
 * @returns Middleware handler
 */
export function validator<
  Type,
  Context,
  Target extends keyof ValidationTargets,
  E extends Env,
  P extends string,
  In = Schema.Schema<Type, Context, never>["Encoded"],
  Out = Schema.Schema<Type, Context, never>["Type"],
  I extends HonoInput = {
    in: HasUndefined<In> extends true
      ? {
          [K in Target]?: K extends "json"
            ? In
            : HasUndefined<keyof ValidationTargets[K]> extends true
              ? { [K2 in keyof In]?: ValidationTargets[K][K2] }
              : { [K2 in keyof In]: ValidationTargets[K][K2] };
        }
      : {
          [K in Target]: K extends "json"
            ? In
            : HasUndefined<keyof ValidationTargets[K]> extends true
              ? { [K2 in keyof In]?: ValidationTargets[K][K2] }
              : { [K2 in keyof In]: ValidationTargets[K][K2] };
        };
    out: { [K in Target]: Out };
  },
  V extends I = I,
>(
  target: Target,
  schema: Schema.Schema<Type, Context, never>,
): MiddlewareHandler<E, P, V> {
  const middleware = effectValidator(target, schema);

  // @ts-expect-error not typed well
  return Object.assign(middleware, {
    [uniqueSymbol]: {
      resolver: async (config: OpenAPIRouteHandlerConfig) =>
        generateValidatorDocs(target, await resolver(schema).builder(config)),
    },
  });
}
