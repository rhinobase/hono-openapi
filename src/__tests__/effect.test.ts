import { Schema } from "effect";
import { Hono } from "hono";
import { describe, expect, it } from "vitest";
import { generateSpecs } from "../handler.js";
import { describeRoute, resolver, validator } from "../middlewares.js";
import { defineCommonTestRouteSpec } from "./utils/route.js";

describe("effect", () => {
  it("basic", async () => {
    const app = new Hono().get(
      "/:id",
      describeRoute(
        defineCommonTestRouteSpec(
          resolver(
            Schema.standardSchemaV1(
              Schema.Struct({
                message: Schema.String,
              }),
            ),
          ),
        ),
      ),
      validator(
        "param",
        Schema.standardSchemaV1(Schema.Struct({ id: Schema.String })),
      ),
      validator(
        "query",
        Schema.standardSchemaV1(
          Schema.Struct({
            category: Schema.Union(
              Schema.Literal("one"),
              Schema.Literal("two"),
              Schema.Literal("three"),
            ).pipe(Schema.optional),
          }),
        ),
      ),
      validator(
        "json",
        Schema.standardSchemaV1(
          Schema.Struct({
            message: Schema.String,
          }),
        ),
      ),
      async (c) => {
        return c.json({ message: "Hello, world!" });
      },
    );

    const specs = await generateSpecs(app);

    expect(specs).toMatchSnapshot();
  });
});
