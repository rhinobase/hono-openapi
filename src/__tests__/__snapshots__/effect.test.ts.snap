// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`effect > basic 1`] = `
{
  "components": {},
  "info": {
    "description": "Development documentation",
    "title": "Hono Documentation",
    "version": "0.0.0",
  },
  "openapi": "3.1.0",
  "paths": {
    "/": {
      "get": {
        "description": "This is a test route",
        "operationId": "GETIndex",
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "additionalProperties": false,
                "properties": {
                  "message": {
                    "type": "string",
                  },
                },
                "required": [
                  "message",
                ],
                "type": "object",
              },
            },
          },
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "additionalProperties": false,
                  "properties": {
                    "message": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "message",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Success",
          },
        },
        "summary": "Test route",
        "tags": [
          "test",
        ],
      },
    },
  },
  "tags": undefined,
}
`;
